src/common/imagwebp.cpp

```diff
@@ -0,0 +1,231 @@
+    if (status != VP8_STATUS_OK)
+    {
+        if (verbose)
+        {
+           wxLogError("WebP: GetFeatures not OK.");
+        }
+        return false;
+    }
+    image->Create(features.width, features.height, false); // this allocates memory
+    if (features.has_alpha)
+    {
+        // image has alpha channel. needs to be decoded, then re-ordered.
+        uint8_t * rgba = WebPDecodeRGBA(webp_data->bytes, webp_data->size, &features.width, &features.height);
+        if (nullptr == rgba)
+        {
+            if (verbose)
+            {
+               wxLogError("WebP: WebPDecodeRGBA failed.");
+            }
+            return false;
+        }
+        image->InitAlpha();
+        unsigned char * rgb = image->GetData();
+        unsigned char * alpha = image->GetAlpha();
+        size_t rgba_index = 0, rgb_index = 0, alpha_index = 0;
+        for (int pixel_counter = 0; pixel_counter < image->GetWidth() * image->GetHeight(); pixel_counter++)
+        {
+            rgb[rgb_index++] = rgba[rgba_index++]; // R
+            rgb[rgb_index++] = rgba[rgba_index++]; // G
+            rgb[rgb_index++] = rgba[rgba_index++]; // B
+            alpha[alpha_index++] = rgba[rgba_index++]; // A
+        }
+        WebPFree(rgba);
+    }
+    else
+    {
+        // image has no alpha channel. decode into target buffer directly.
+        int buffer_size = image->GetWidth() * image->GetHeight() * 3;
+        int stride = image->GetWidth() * 3;
+        uint8_t * output_buffer = WebPDecodeRGBInto(webp_data->bytes, webp_data->size, image->GetData(), buffer_size, stride);
+        if (nullptr == output_buffer)
+        {
+            if (verbose)
+            {
+               wxLogError("WebP: WebPDecodeRGBInto failed.");
+            }
+            return false;
+        }
+    }
+    image->SetMask(false); // all examples do this, so I do so as well
+    return true;
+}
+
+bool DecodeWebPFrameIntoImage(wxImage *image, int index, WebPDemuxerPtr & demuxer, bool verbose)
+{
+    bool ok = false;
+    // apparently, index can be -1 for "don't care", but libwebp does care
+    if (index < 0)
+    {
+        index = 0;
+    }
+    WebPIterator iter;
+    // wxImageHandler index starts from 0 (first frame)
+    // WebPDemuxGetFrame to starts from 1 (0 means "last frame")
+    if (WebPDemuxGetFrame(demuxer.get(), index+1, &iter))
+    {
+        ok = DecodeWebPDataIntoImage(image, &iter.fragment, verbose);
+        WebPDemuxReleaseIterator(&iter);
+    }
+    return ok;
+}
+
+WebPDemuxerPtr CreateDemuxer(wxInputStream& stream, bool verbose = false) {
+    wxMemoryOutputStream * mos = new wxMemoryOutputStream;
+    stream.Read(*mos); // this reads the entire file into memory
+    // TODO: Only read data as needed. WebPDemux can operate on partial data.
+    //       Could save some bandwidth with e.g. DoGetImageCount
+    wxStreamBuffer * mosb = mos->GetOutputStreamBuffer();
+    WebPData * webp_data = new WebPData;
```

Comment:
Does this have to live for as long as `WebPDemuxer` itself lives or can it be destroyed immediately after calling `WebPDemux()` (as I suspect)?

In the latter case, we don't need to create it (nor `mos`) on the heap, which avoids the need to `delete` it too. In the former, I'd strongly prefer to keep it in a single object with `wxMemoryOutputStream` and `WebPDemuxer*` itself rather than hiding them, which are logically part of the same object, inside a custom deleter.

Reply:
> Does this have to live for as long as WebPDemuxer itself lives

I assume, with "this" you refer to the `wxMemoryOutputStream` `mos` is pointing to. As far as I understood, `WebPDemux` does _not_ create a copy of the input data in all cases. At least that is what I make of the implementation of [WebPDemuxInternal](https://chromium.googlesource.com/webm/libwebp/+/refs/tags/v1.4.0/src/demux/demux.c#705). Unfortunately, the documentation is not terribly clear about the expected life-time of the buffers.

To keep the implementation simple for now, I went with the "read entire file into memory" approach. I hope that is okay. Is`wxInputStream::read()` into a `wxMemoryOutputStream` the way it is done in wxWidgets?

> I'd strongly prefer to keep it in a single object

I can do that.

Reply:
I meant both `WebPData` itself and the buffer associated with it.

Looking at the source (thanks for the link), it seems like `data` is only used inside this function, as it's used to initialize the local `MemBuffer` variable and so can't be accessed once the function returns. So it really seems like we don't need these heap allocations (and corresponding `delete`s) and can just make them local variables to simplify things.

Reply:
I now remember (sorry, it has been almost half a year, the details are a bit hazy) that I need the `WebPDemux` in `DoGetImageCount`, too. I did not want redundant code for creating a `WebPDemux` struct from a `wxInputStream`. I do not recall whether `DoGetImageCount` works for me or not, but I did not take the time to create any tests. My comment indicates that is the only reason why `DoGetImageCount` is commented out.

As far as I understand, `data` – or rather the buffer associated with it – is also used later on: The demuxer emits `WebPData` structs pointing to different sections within `data`. Instances include the call to `WebPGetFeatures` for getting the image dimensions. Also the actual decompression with `WebPDecodeRGB…`.

Indeed, it looks like we do not need the heap-allocated `WebPData` from line 123 after `WebPDemux()`. However, I am pretty confident that `data` (aka `wxMemoryOutputStream mos`) must remain valid during `WebPDemux`'s lifetime. I should verify that when touching the code again.